From: Ursula Braun <ubraun@linux.vnet.ibm.com>
Date: Mon, 4 Jul 2016 14:07:16 +0200
Subject: qeth: delete napi struct when removing a qeth device
Git-commit: 7831b4ff0d926e0deeaabef9db8800ed069a2757
Patch-mainline: v4.7-rc7
References: bnc#979915, LTC#143590

Description:  qeth: delete napi struct when removing a qeth device
Symptom:      oops in napi_hash_add() called from __qeth_l2_set_online()
Problem:      A qeth_card contains a napi_struct linked to the net_device
              during device probing. This struct must be deleted when removing
              the qeth device.
Solution:     add the missing netif_napi_del() calls
Reproduction: generate heavy network load, repeatedly toogle OSA chpids to
              remove and add qeth devices

Upstream-Description:

              qeth: delete napi struct when removing a qeth device

              A qeth_card contains a napi_struct linked to the net_device during
              device probing. This struct must be deleted when removing the qeth
              device, otherwise Panic on oops can occur when qeth devices are
              repeatedly removed and added.

              Fixes: a1c3ed4c9ca ("qeth: NAPI support for l2 and l3 discipline")
              Cc: stable@vger.kernel.org # v2.6.37+
              Signed-off-by: Ursula Braun <ubraun@linux.vnet.ibm.com>
              Tested-by: Alexander Klein <ALKL@de.ibm.com>
              Signed-off-by: David S. Miller <davem@davemloft.net>


Signed-off-by: Ursula Braun <ubraun@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/net/qeth_l2_main.c |    1 +
 drivers/s390/net/qeth_l3_main.c |    1 +
 2 files changed, 2 insertions(+)

--- a/drivers/s390/net/qeth_l2_main.c
+++ b/drivers/s390/net/qeth_l2_main.c
@@ -898,6 +898,7 @@ static void qeth_l2_remove_device(struct
 		qeth_l2_set_offline(cgdev);
 
 	if (card->dev) {
+		netif_napi_del(&card->napi);
 		unregister_netdev(card->dev);
 		card->dev = NULL;
 	}
--- a/drivers/s390/net/qeth_l3_main.c
+++ b/drivers/s390/net/qeth_l3_main.c
@@ -3396,6 +3396,7 @@ static void qeth_l3_remove_device(struct
 		qeth_l3_set_offline(cgdev);
 
 	if (card->dev) {
+		netif_napi_del(&card->napi);
 		unregister_netdev(card->dev);
 		card->dev = NULL;
 	}
